### PR TITLE
fix(tracing): respect OTEL_TRACES_SAMPLER env var in NewOTelFromEnv

### LINKED ISSUE #931: `NewOTelFromEnv` ignores `OTEL_TRACES_SAMPLER` by hardcoding `AlwaysSample()`
`NewOTelFromEnv` in [tracing/otel.go](https://github.com/grafana/dskit/blob/main/tracing/otel.go#L37) hardcodes `tracesdk.AlwaysSample()` in the else branch (when no Jaeger remote sampler is configured), which overrides any sampler the OTel SDK would have configured from the standard `OTEL_TRACES_SAMPLER` environment variable.

The current code:

```go
options = append(options, tracesdk.WithSampler(
    &JaegerDebuggingSampler{tracesdk.ParentBased(tracesdk.AlwaysSample())}))
```

The comment in `maybeJaegerRemoteSamplerFromEnv` states *"trace provider already configured it through samplerFromEnv()"*, but the explicit `WithSampler(AlwaysSample)` call immediately overrides whatever the SDK configured.

## Impact

Any dskit-based service (Mimir, Loki, Tempo, etc.) that sets `OTEL_TRACES_SAMPLER=traceidratio` with `OTEL_TRACES_SAMPLER_ARG=0.1` will still sample 100% of traces. This makes it impossible to control trace volume through the standard OTel environment variables, which can cause significant overhead in production.

## Steps to reproduce

1. Run Mimir (or any dskit-based service) with:
   ```
   OTEL_EXPORTER_OTLP_ENDPOINT=http://collector:4317
   OTEL_TRACES_SAMPLER=traceidratio
   OTEL_TRACES_SAMPLER_ARG=0.1
   ```
2. Observe that 100% of traces are exported, not 10%.

## Expected behavior

The sampler should respect `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` as defined in the [OpenTelemetry specification](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler).

## Suggested fix

Replace `tracesdk.ParentBased(tracesdk.AlwaysSample())` with a helper that reads the standard env vars:

```go
func otelSamplerFromEnv() tracesdk.Sampler {
    samplerName := os.Getenv("OTEL_TRACES_SAMPLER")
    samplerArg  := os.Getenv("OTEL_TRACES_SAMPLER_ARG")

    switch samplerName {
    case "always_on":
        return tracesdk.ParentBased(tracesdk.AlwaysSample())
    case "always_off":
        return tracesdk.ParentBased(tracesdk.NeverSample())
    case "traceidratio":
        return tracesdk.ParentBased(tracesdk.TraceIDRatioBased(parseRatio(samplerArg)))
    case "parentbased_always_on":
        return tracesdk.ParentBased(tracesdk.AlwaysSample())
    case "parentbased_always_off":
        return tracesdk.ParentBased(tracesdk.NeverSample())
    case "parentbased_traceidratio":
        return tracesdk.ParentBased(tracesdk.TraceIDRatioBased(parseRatio(samplerArg)))
    default:
        return tracesdk.ParentBased(tracesdk.AlwaysSample())
    }
}
```

Then in `NewOTelFromEnv`:
```go
options = append(options, tracesdk.WithSampler(
    &JaegerDebuggingSampler{otelSamplerFromEnv()}))
```

## Affected version

Confirmed on dskit commit `a4bfa505f5e9` (vendored by Mimir 3.0.2).

### PR DESCRIPTION
**Pull request by bubu11e**
## Problem

  `NewOTelFromEnv` hardcodes `AlwaysSample()` in its else branch (when no Jaeger
  remote sampler is configured), which overrides any sampler the OTel SDK would
  have configured from the standard `OTEL_TRACES_SAMPLER` environment variable:

      options = append(options, tracesdk.WithSampler(
          &JaegerDebuggingSampler{tracesdk.ParentBased(tracesdk.AlwaysSample())}))

  The comment in `maybeJaegerRemoteSamplerFromEnv` states that the "trace provider
  already configured it through samplerFromEnv()", but the explicit `WithSampler`
  call immediately overrides it. As a result, any dskit-based service setting
  `OTEL_TRACES_SAMPLER=traceidratio` with `OTEL_TRACES_SAMPLER_ARG=0.1` still
  samples 100 % of traces, making it impossible to control trace volume through
  standard OTel environment variables.

  ## Fix

  Replace the hardcoded sampler with `otelSamplerFromEnv()`, a helper that reads
  `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` and returns the correct
  sampler per the [OTel specification][spec]. When the env vars are unset, it
  defaults to `ParentBased(AlwaysSample())`, preserving backward compatibility.

  Supported sampler values:

  | `OTEL_TRACES_SAMPLER`      | Sampler                                 |
  |----------------------------|-----------------------------------------|
  | `always_on`                | `AlwaysSample()`                        |
  | `always_off`               | `NeverSample()`                         |
  | `traceidratio`             | `TraceIDRatioBased(arg)`                |
  | `parentbased_always_on`    | `ParentBased(AlwaysSample())`           |
  | `parentbased_always_off`   | `ParentBased(NeverSample())`            |
  | `parentbased_traceidratio` | `ParentBased(TraceIDRatioBased(arg))`   |

  [spec]: https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler

## Unit tests

  - TestOtelSamplerFromEnv — 12 sub-tests covering every sampler type plus
  invalid/missing args and out-of-range ratios
  - TestParseRatioOrDefault — 7 sub-tests for ratio parsing edge cases
  - TestNewOTelFromEnvRespectsSampler — regression test confirming
  traceidratio with 0.1 produces TraceIDRatioBased{0.1}, not
  AlwaysOnSampler

## End-to-end validation with Mimir

The fix can be verified end-to-end by building Mimir against the patched dskit
and checking that OTEL_TRACES_SAMPLER actually controls trace output.

The test stack uses four containers on an internal Docker network:

Prometheus ──scrape + remote write──> Mimir ──OTLP traces──> Alloy ──> Tempo

Prometheus scrapes Mimir metrics and remote-writes them back, which generates
internal Mimir trace activity. Alloy collects the OTLP spans and forwards them
to Tempo.

Directory layout

dskit-931-test/
- docker-compose.yml
- mimir/
    - Dockerfile
    - config.yaml
- alloy/
    - config.alloy
- tempo/
    - tempo.yaml
- prometheus/
    - prometheus.yml

### Files
docker-compose.yml:
```yaml
services:
  mimir:
    build:
      context: ./mimir
      dockerfile: Dockerfile
    container_name: mimir
    volumes:
      - ./mimir/config.yaml:/etc/mimir/config.yaml
    command: ["-config.file=/etc/mimir/config.yaml", "-target=all"]
    environment:
      - OTEL_EXPORTER_OTLP_ENDPOINT=http://alloy:4317
      - OTEL_EXPORTER_OTLP_PROTOCOL=grpc
      - OTEL_SERVICE_NAME=mimir
      - OTEL_TRACES_SAMPLER=traceidratio
      - OTEL_TRACES_SAMPLER_ARG=0.5
    networks: [test]
    depends_on: [alloy]

  alloy:
    image: grafana/alloy:v1.13.0
    container_name: alloy
    volumes:
      - ./alloy/config.alloy:/etc/alloy/config.alloy
    command: [run, --server.http.listen-addr=0.0.0.0:12345, /etc/alloy/config.alloy]
    networks: [test]
    depends_on: [tempo]

  tempo:
    image: grafana/tempo:2.10.0
    container_name: tempo
    volumes:
      - ./tempo/tempo.yaml:/etc/tempo.yaml
    command: ["--config.file=/etc/tempo.yaml"]
    networks: [test]

  prometheus:
    image: prom/prometheus:v3.9.1
    container_name: prometheus
    volumes:
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
    networks: [test]

networks:
  test:
```

mimir/Dockerfile:
```Dockerfile
ARG GO_VERSION=1.25
ARG MIMIR_VERSION=mimir-3.0.2
ARG DSKIT_COMMIT=a4bfa505f5e9

FROM golang:${GO_VERSION}-bookworm AS builder

ARG MIMIR_VERSION
ARG DSKIT_COMMIT

RUN apt-get update && apt-get install -y --no-install-recommends git && rm -rf /var/lib/apt/lists/*

WORKDIR /src

# Clone upstream dskit at the Mimir-vendored commit, then cherry-pick the fix
# from the PR branch (bubu11e/dskit#fix/otel-sampler-from-env → grafana/dskit#944)
RUN git clone https://github.com/grafana/dskit.git /src/dskit && \
    cd /src/dskit && \
    git checkout ${DSKIT_COMMIT} && \
    git remote add fix https://github.com/bubu11e/dskit.git && \
    git fetch fix fix/otel-sampler-from-env && \
    git cherry-pick --no-commit fix/fix/otel-sampler-from-env

RUN git clone --depth 1 --branch ${MIMIR_VERSION} \
    https://github.com/grafana/mimir.git /src/mimir

WORKDIR /src/mimir
RUN go mod edit -replace github.com/grafana/dskit=/src/dskit && \
    go mod tidy && go mod vendor
RUN CGO_ENABLED=0 go build -o /bin/mimir ./cmd/mimir

# Minimal runtime
FROM gcr.io/distroless/static-debian12:nonroot
COPY --from=builder /bin/mimir /bin/mimir
ENTRYPOINT ["/bin/mimir"]
```

mimir/config.yaml:
```yaml
multitenancy_enabled: false

server:
  http_listen_address: 0.0.0.0
  grpc_listen_port: 9009
  http_listen_port: 8080
  log_format: json
  log_level: warn

distributor:
  ring:
    kvstore:
      store: inmemory

ingester:
  ring:
    kvstore:
      store: inmemory
    replication_factor: 1
```
alloy/config.alloy:
```json
// OTLP receiver — Mimir sends traces here.
otelcol.receiver.otlp "default" {
  grpc { endpoint = "0.0.0.0:4317" }
  http { endpoint = "0.0.0.0:4318" }

  output {
    traces = [otelcol.processor.batch.default.input]
  }
}

otelcol.processor.batch "default" {
  timeout = "1s"

  output {
    traces = [otelcol.exporter.otlp.tempo.input]
  }
}

// Forward traces to Tempo via OTLP gRPC.
otelcol.exporter.otlp "tempo" {
  client {
    endpoint = "tempo:4317"
    tls { insecure = true }
  }
}
```
tempo/tempo.yaml
```yaml
server:
  http_listen_port: 3200

distributor:
  receivers:
    otlp:
      protocols:
        grpc:
          endpoint: 0.0.0.0:4317
        http:
          endpoint: 0.0.0.0:4318

ingester:
  max_block_duration: 5m

compactor:
  compaction:
    block_retention: 1h

storage:
  trace:
    backend: local
    wal:
      path: /var/tempo/wal
    local:
      path: /var/tempo/blocks
```
prometheus/prometheus.yml
```yaml
scrape_configs:
  - job_name: mimir
    static_configs:
      - targets: ['mimir:8080']

remote_write:
  - url: http://mimir:8080/api/v1/push
```

### Running the test
```bash
# Build and start the stack
docker compose up -d --build

# Wait ~30s for Prometheus to start remote-writing and Mimir to generate traces

# Record current span count at the collector
BEFORE=$(docker compose exec prometheus \
  wget -qO- 'http://alloy:12345/metrics' | \
  grep 'otelcol_receiver_accepted_spans_total.*grpc' | grep -oP '\d+$')

sleep 30

AFTER=$(docker compose exec prometheus \
  wget -qO- 'http://alloy:12345/metrics' | \
  grep 'otelcol_receiver_accepted_spans_total.*grpc' | grep -oP '\d+$')

echo "New spans with traceidratio(0.5): $((AFTER - BEFORE))"
# Expected: > 0 (traces are flowing, sampled at 50%)
```

Then switch to always_off to prove the sampler env var is actually respected:
```bash
# Change OTEL_TRACES_SAMPLER=always_off in docker-compose.yml, then:
docker compose up -d mimir

BEFORE=$(docker compose exec prometheus \
  wget -qO- 'http://alloy:12345/metrics' | \
  grep 'otelcol_receiver_accepted_spans_total.*grpc' | grep -oP '\d+$')

sleep 30

AFTER=$(docker compose exec prometheus \
  wget -qO- 'http://alloy:12345/metrics' | \
  grep 'otelcol_receiver_accepted_spans_total.*grpc' | grep -oP '\d+$')

echo "New spans with always_off: $((AFTER - BEFORE))"
# Expected: 0 (no new spans)
# Before this fix: spans keep flowing regardless of OTEL_TRACES_SAMPLER
```

 ### Observed results

| `OTEL_TRACES_SAMPLER` | New spans (30s) | Before fix |
|------------------------|-----------------|------------|
| `traceidratio` (0.5)  | 11 (sampled)    | ~22 (all)  |
| `always_off`           | **0**           | ~22 (all)  |

  ### Clean up
  docker compose down

  Fixes #931

---

> [!NOTE]
> **Medium Risk**
> Changes default trace sampling behavior to honor env-configured samplers, which can materially alter trace volume and operational visibility. Scope is limited to tracing init and is covered by targeted unit/regression tests.
> 
> **Overview**
> `NewOTelFromEnv` no longer hardcodes `ParentBased(AlwaysSample())` when no Jaeger remote sampler is configured; it now derives the sampler from `OTEL_TRACES_SAMPLER`/`OTEL_TRACES_SAMPLER_ARG` (while still wrapping with `JaegerDebuggingSampler`).
> 
> This adds `otelSamplerFromEnv` (plus spec-aligned constants and ratio parsing/clamping) and normalizes Jaeger remote sampler env parsing, with new unit/regression tests covering supported sampler modes and invalid args.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 82f4457cfdda9a8d0e1a98b59a9493cb8ff0f6ba. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tracing/otel.go [hunk 1]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+
+// parseRatioOrDefault parses a float64 from s, clamped to [0, 1].
+// Returns defaultVal if s is empty, unparseable, or out of range.
+func parseRatioOrDefault(s string, defaultVal float64) float64 {
+	if s == "" {
+		return defaultVal
+	}
+	v, err := strconv.ParseFloat(s, 64)
+	if err != nil || v < 0 || v > 1 {
+		return defaultVal
```
[the remark this answers is not in the document]
Comment (bubu11e) [reply tracing/otel.go:330]: Fixed, A test have been added.

**Comment by pracucci** [comment]
Hi! 👋 We recently merged #947 which changed the CI configuration for unit tests. To get CI passing on this PR, please rebase on the latest `main` branch:

```
git fetch origin main
git rebase origin/main
git push --force-with-lease
```

Thanks!

**Comment by bubu11e** [comment]
> Hi! 👋 We recently merged #947 which changed the CI configuration for unit tests. To get CI passing on this PR, please rebase on the latest `main` branch:
> 
> ```
> git fetch origin main
> git rebase origin/main
> git push --force-with-lease
> ```
> 
> Thanks!

Hi, rebase is done.

File: tracing/otel.go [hunk 2]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+func otelSamplerFromEnv() tracesdk.Sampler {
+	samplerName := os.Getenv("OTEL_TRACES_SAMPLER")
+	samplerArg := os.Getenv("OTEL_TRACES_SAMPLER_ARG")
+
+	switch samplerName {
+	case "always_on":
+		return tracesdk.AlwaysSample()
+	case "always_off":
+		return tracesdk.NeverSample()
+	case "traceidratio":
```
Comment (bboreham) [inline tracing/otel.go]: Do we really have to implement this ourselves?
No utility function in Otel that does it?

Comment (bubu11e) [reply tracing/otel.go]: > Do we really have to implement this ourselves? No utility function in Otel that does it?

The OTel Go SDK has this logic internally ([sampler_env.go](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/sampler_env.go#L48)) but it's unexported — I did not find a public SamplerFromEnv() we can call. So I'm afraid, we do have to implement the switch ourselves.

File: tracing/otel.go [hunk 3]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+		return tracesdk.ParentBased(tracesdk.NeverSample())
+	case "parentbased_traceidratio":
+		return tracesdk.ParentBased(tracesdk.TraceIDRatioBased(parseRatioOrDefault(samplerArg, 1.0)))
+	default:
+		// Matches the upstream default when no sampler is configured.
+		return tracesdk.ParentBased(tracesdk.AlwaysSample())
+	}
+}
+
+// parseRatioOrDefault parses a float64 from s, clamped to [0, 1].
```
Comment (javiermolinar) [inline tracing/otel.go]:
```suggestion
// parseRatioOrDefault parses a float64 from s, clamped to [0, defaultVal].
```

File: tracing/otel.go [hunk 4]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+//
+// Supported values per the OpenTelemetry specification:
+//   - "always_on"                  → AlwaysSample
+//   - "always_off"                 → NeverSample
+//   - "traceidratio"               → TraceIDRatioBased(arg)
+//   - "parentbased_always_on"      → ParentBased(AlwaysSample)
+//   - "parentbased_always_off"     → ParentBased(NeverSample)
+//   - "parentbased_traceidratio"   → ParentBased(TraceIDRatioBased(arg))
+//
+// See https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler
```
Comment (javiermolinar) [inline tracing/otel.go:297]: Could you add here the link to the source function? So we can update it if something changes upstream

`https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/sampler_env.go#L48`

File: tracing/otel.go [hunk 5]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+func otelSamplerFromEnv() tracesdk.Sampler {
+	samplerName := os.Getenv("OTEL_TRACES_SAMPLER")
+	samplerArg := os.Getenv("OTEL_TRACES_SAMPLER_ARG")
+
+	switch samplerName {
+	case "always_on":
```
Comment (javiermolinar) [inline tracing/otel.go]: wdyt if we add a constant block as the otel reference

```go
const (
	tracesSamplerKey    = "OTEL_TRACES_SAMPLER"
	tracesSamplerArgKey = "OTEL_TRACES_SAMPLER_ARG"

	samplerAlwaysOn                = "always_on"
	samplerAlwaysOff               = "always_off"
	samplerTraceIDRatio            = "traceidratio"
	samplerParentBasedAlwaysOn     = "parentbased_always_on"
	samplerParsedBasedAlwaysOff    = "parentbased_always_off"
	samplerParentBasedTraceIDRatio = "parentbased_traceidratio"
)
```

**Review by javiermolinar — COMMENTED** [review]
A couple of nit but LGTM

File: tracing/otel.go [hunk 6]
```diff
@@ -269,6 +267,55 @@ type closableParentBasedSampler struct {
...
+//   - "always_on"                  → AlwaysSample
+//   - "always_off"                 → NeverSample
+//   - "traceidratio"               → TraceIDRatioBased(arg)
+//   - "parentbased_always_on"      → ParentBased(AlwaysSample)
+//   - "parentbased_always_off"     → ParentBased(NeverSample)
+//   - "parentbased_traceidratio"   → ParentBased(TraceIDRatioBased(arg))
+//
+// See https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler
+func otelSamplerFromEnv() tracesdk.Sampler {
+	samplerName := os.Getenv("OTEL_TRACES_SAMPLER")
```
Comment (javiermolinar) [inline tracing/otel.go]: We should normalize the env vars here as well to support upper case

```go
sampler = strings.ToLower(strings.TrimSpace(sampler))
samplerArg = strings.TrimSpace(samplerArg)
```

File: tracing/otel.go (hunk 3 above)
Comment (bubu11e) [reply tracing/otel.go]: Thanks, done.

File: tracing/otel.go (hunk 5 above)
Comment (bubu11e) [reply tracing/otel.go]: Thanks, done.

File: tracing/otel.go (hunk 6 above)
Comment (bubu11e) [reply tracing/otel.go]: Thanks, done.

File: tracing/otel.go (hunk 4 above)
Comment (bubu11e) [reply tracing/otel.go:297]: Thanks, done.

**Review by javiermolinar — APPROVED** [review]
Looking good, thank you for your contribution!
